shell/platform/android/platform_view_android.cc

```diff
@@ -229,27 +229,51 @@ void PlatformViewAndroid::SetViewportMetrics(JNIEnv* env,
   });
 }
 
+//void PlatformViewAndroid::DispatchPlatformMessage(JNIEnv* env,
+//                                                  jobject obj,
+//                                                  jstring java_name,
+//                                                  jstring java_message_data,
+//                                                  jint response_id) {
+//  std::string name = base::android::ConvertJavaStringToUTF8(env, java_name);
+//  std::string data;
+//  if (java_message_data)
+//    data = base::android::ConvertJavaStringToUTF8(env, java_message_data);
+//
+//  ftl::RefPtr<blink::PlatformMessageResponse> response;
+//  if (response_id) {
+//    response = ftl::MakeRefCounted<PlatformMessageResponseAndroid>(
+//        response_id, GetWeakPtr());
+//  }
+//
+//  const uint8_t* buffer = reinterpret_cast<const uint8_t*>(data.data());
+//  PlatformView::DispatchPlatformMessage(
+//      ftl::MakeRefCounted<blink::PlatformMessage>(
+//          std::move(name), std::vector<uint8_t>(buffer, buffer + data.size()),
+//          std::move(response)));
+//}
 void PlatformViewAndroid::DispatchPlatformMessage(JNIEnv* env,
                                                   jobject obj,
                                                   jstring java_name,
-                                                  jstring java_message_data,
+                                                  jobject java_message_data,
+                                                  jint java_message_position,
                                                   jint response_id) {
   std::string name = base::android::ConvertJavaStringToUTF8(env, java_name);
-  std::string data;
-  if (java_message_data)
-    data = base::android::ConvertJavaStringToUTF8(env, java_message_data);
+  std::vector<uint8_t> message;
+  if (java_message_data) {
+    uint8_t* message_data = static_cast<uint8_t*>(env->GetDirectBufferAddress(java_message_data));
+    message = std::vector<uint8_t>(message_data, message_data + java_message_position);
+  }
+
 
   ftl::RefPtr<blink::PlatformMessageResponse> response;
   if (response_id) {
     response = ftl::MakeRefCounted<PlatformMessageResponseAndroid>(
         response_id, GetWeakPtr());
   }
 
-  const uint8_t* buffer = reinterpret_cast<const uint8_t*>(data.data());
   PlatformView::DispatchPlatformMessage(
-      ftl::MakeRefCounted<blink::PlatformMessage>(
-          std::move(name), std::vector<uint8_t>(buffer, buffer + data.size()),
-          std::move(response)));
+      ftl::MakeRefCounted<blink::PlatformMessage>(std::move(name), message,
```

Comment:
You want to `std::move` the `message` too to avoid a memcpy of the message payload.